addons/mail/static/src/rtc/voice_recorder.js

```diff
@@ -0,0 +1,51 @@
+/* @odoo-module */
+import { Component, useState, useRef } from "@odoo/owl";
+import { useService } from "@web/core/utils/hooks";
+import { useAttachmentUploader } from "@mail/attachments/attachment_uploader_hook";
+
+export class VoiceRecorder extends Component {
+    static props = ["thread", "composer?"];
+    static template = "mail.VoiceRecorder";
+
+    setup() {
+        this.audioRef = useRef("audio");
+        this.state = useState({
+            recording: false,
```

Comment:
It's strange because the `recording` state is at the component level, but the rest of the recording is at the service level. I don't know how many composer we can have at the same time but you should probably at least stop the recording when the composer is destroyed.

Reply:
The service is overkill. In the additional commit I pushed, I suggest to just put everything in the component. The component should handle killing resources itself.